### PR TITLE
Test invalid builder rates

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- expand push queue builder docs with a complete example covering capacities, rate limits, and dead-letter queue behaviour
- document rate and DLQ setters, noting that `None` disables throttling and frames are dropped when the DLQ is absent or full
- clarify lint expectations in connection actor fixtures and validate builder rates with a dedicated `is_invalid_rate` predicate
- refine deprecated constructors' error messages and mark `PushError` as non-exhaustive
- deduplicate deprecated constructors via a shared `build_via_builder` helper
- reserve queue slots before rate limiting and throttle DLQ logging with per-handle counters
- test that `rate(None)` disables throttling and share builder fixtures across push policy tests
- consolidate test queue construction with fixtures to reduce repetition
- replace spaced hyphens with em dashes, mark docs `no_run`, disable throttling in more tests, remove `#[tokio::test]` from doctests, and downgrade DLQ loss logs to warnings
- adopt Oxford “-ize” spelling across documentation and swap unchecked `unwrap` calls for `expect` in push handle examples

## Testing
- `make fmt`
- `make lint`
- `make test`
- `make markdownlint`
- `make nixie` *(fails: KeyboardInterrupt)*

------
https://chatgpt.com/codex/tasks/task_e_68aeba8d630c8322bcab55de80b2ef06

## Summary by Sourcery

Replace PushQueues’ legacy constructors with a flexible PushQueuesBuilder API and enforce robust rate‐limit and dead‐letter queue behavior across code, tests, and documentation

New Features:
- Introduce PushQueuesBuilder for fluent configuration of high/low capacities, rate limits, and optional DLQ
- Add dead‐letter queue support and make rate limiting optional via None

Bug Fixes:
- Fix builder to reject invalid rates with a dedicated is_invalid_rate predicate

Enhancements:
- Enforce and validate allowed push rates (reject zero or above MAX_PUSH_RATE, accept None for unlimited)
- Reserve queue slots before applying rate limiting and throttle DLQ logging with per‐handle counters
- Deprecate old constructors in favor of builder and mark PushError and PushConfigError as non‐exhaustive
- Downgrade DLQ loss logs to warnings

Documentation:
- Expand PushQueues documentation with example covering capacities, rate limits, and DLQ behavior and mark code blocks no_run
- Adopt Oxford “-ize” spelling, replace spaced hyphens with em dashes, and replace unwraps with expect in examples

Tests:
- Add tests for invalid builder rates, unlimited mode, and capacity validations
- Unify and deduplicate push queue fixtures and share builder setup across push and policy tests

Chores:
- Remove #[tokio::test] from doctests, apply fmt/lint/markdownlint fixes

### CODE REVIEW

File: src/push/queues/handle.rs [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+            .clone()
+            .reserve_owned()
+            .await
+            .map_err(|_| PushError::Closed)?;
+
+        if let Some(ref limiter) = self.0.limiter {
+            limiter.acquire(1).await;
+        }
+
+        permit.send(frame);
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/queues/handle.rs:65]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

File: tests/push.rs [hunk 2]
```diff
@@ -2,15 +2,73 @@
...
+    let result = PushQueues::<u8>::builder().rate(Some(rate)).build();
+    assert!(matches!(result, Err(PushConfigError::InvalidRate(r)) if r == rate));
+}
+
+/// Setting the rate to `None` disables throttling without error.
+#[test]
+fn builder_accepts_none_rate() {
+    let result = PushQueues::<u8>::builder().rate(None).build();
+    assert!(result.is_ok(), "builder should accept None rate");
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/push.rs]: Has this now been resolved?

File: tests/push.rs [hunk 3]
```diff
@@ -2,15 +2,73 @@
...
+
+#[test]
+fn builder_rejects_zero_capacity() {
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/push.rs:82]: Has this now been resolved?

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

- Add a test case for rate = MAX_PUSH_RATE to verify that the builder accepts the upper boundary as valid.
- There’s a lot of repeated builder boilerplate in the tests—consider extracting a shared helper or fixture for default queue construction to reduce duplication.
- Consider adding a convenience builder method like `unlimited()` that wraps `.rate(None)` to make the no-rate-limiting scenario more ergonomic.

File: src/push/queues/handle.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/queues/handle.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: src/push/queues/builder.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        self
+    }
+
+    /// Build the configured [`PushQueues`] and associated [`PushHandle`].
+    ///
+    /// # Errors
+    ///
+    /// Returns [`PushConfigError::InvalidRate`] if the rate is zero or strictly
+    /// greater than [`super::MAX_PUSH_RATE`] and
+    /// [`PushConfigError::InvalidCapacity`] if either queue capacity is zero.
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/queues/builder.rs]: Has this now been resolved?

**Comment by leynos** [comment]
@coderabbitai review

File: src/push/queues/mod.rs [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+    ) -> Self {
+        Self {
+            high_capacity,
+            low_capacity,
+            rate,
+            dlq,
+            dlq_log_every_n: 100,
+            dlq_log_interval: Duration::from_secs(10),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/queues/mod.rs:99]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

**Comment by leynos** [comment]
@coderabbitai pause

File: src/push/mod.rs [hunk 6]
```diff
@@ -0,0 +1,31 @@
+//! Prioritised queues used for asynchronously pushing frames to a connection.
+//!
+//! # Overview
+//! Expose prioritised push queues and their handle. Construct via a fluent
+//! builder.
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/mod.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: src/push/queues/builder.rs [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    pub fn build(self) -> Result<(PushQueues<F>, PushHandle<F>), PushConfigError> {
+        PushQueues::build_with_rate_dlq(
+            self.high_capacity,
+            self.low_capacity,
+            self.rate,
+            self.dlq,
+            self.dlq_log_every_n,
+            self.dlq_log_interval,
+        )
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/queues/builder.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: src/push/queues/mod.rs [hunk 8]
```diff
@@ -0,0 +1,279 @@
+//! Queue management used by [`PushHandle`] and [`PushQueues`].
+//!
+//! Provides the core implementation for prioritised queues delivering frames
+//! to a connection. Background tasks can send messages without blocking the
+//! request/response cycle. Frames maintain FIFO order within each priority
+//! level. An optional rate limiter caps throughput at [`MAX_PUSH_RATE`] pushes
+//! per second.
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/queues/mod.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: src/push/queues/mod.rs [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+    ///     let (mut queues, handle) = PushQueues::<u8>::builder()
+    ///         .high_capacity(1)
+    ///         .low_capacity(1)
+    ///         .build()
+    ///         .expect("failed to build PushQueues");
+    ///     handle.push_high_priority(2u8).await.expect("push failed");
+    ///     let (priority, frame) = queues.recv().await.expect("recv failed");
+    ///     assert_eq!(priority, PushPriority::High);
+    ///     assert_eq!(frame, 2);
+    /// }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/queues/mod.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: src/session.rs [hunk 10]
```diff
@@ -83,7 +83,9 @@ impl<F: FrameLike> SessionRegistry<F> {
     ///     session::{ConnectionId, SessionRegistry},
     /// };
     ///
-    /// let (_queues, handle) = PushQueues::<u8>::bounded(1, 1);
+    /// let (_queues, handle) = PushQueues::<u8>::builder()
+    ///     .build()
+    ///     .expect("failed to build PushQueues");
     /// let registry = SessionRegistry::default();
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/session.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: tests/connection_actor.rs [hunk 11]
```diff
@@ -18,24 +18,39 @@ use wireframe::{
...
 #[fixture]
-#[allow(
+#[expect(
     unused_braces,
     reason = "rustc false positive for single line rstest fixtures"
 )]
+// allow(unfulfilled_lint_expectations): rustc occasionally fails to emit the expected
+// lint for single-line rstest fixtures on stable.
+#[allow(unfulfilled_lint_expectations)]
 fn shutdown_token() -> CancellationToken { CancellationToken::new() }
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/connection_actor.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: src/push/queues/mod.rs [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+    ///
+    /// # Panics
+    ///
+    /// Panics if an internal invariant is violated. This should never occur.
+    #[deprecated(since = "0.1.0", note = "Use `PushQueues::builder` instead")]
+    #[must_use]
+    pub fn bounded(high_capacity: usize, low_capacity: usize) -> (Self, PushHandle<F>) {
+        Self::build_via_builder(high_capacity, low_capacity, Some(DEFAULT_PUSH_RATE), None)
+            .expect("invalid capacities or rate in deprecated bounded()")
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/queues/mod.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: src/push/queues/mod.rs [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+    ) -> Result<(Self, PushHandle<F>), PushConfigError> {
+        Self::build_with_config(PushQueueConfig {
+            high_capacity,
+            low_capacity,
+            rate,
+            dlq,
+            dlq_log_every_n: 100,
+            dlq_log_interval: Duration::from_secs(10),
+        })
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/queues/mod.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: src/push/queues/mod.rs [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+impl<T> FrameLike for T where T: Send + 'static {}
+
+// Default maximum pushes per second when no custom rate is specified.
+// This is an internal implementation detail and may change.
+const DEFAULT_PUSH_RATE: usize = 100;
+/// Highest supported rate for [`PushQueuesBuilder::rate`].
+pub const MAX_PUSH_RATE: usize = 10_000;
+
+// Compile-time guard: DEFAULT_PUSH_RATE must not exceed MAX_PUSH_RATE.
+const _: () = assert!(DEFAULT_PUSH_RATE <= MAX_PUSH_RATE);
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/queues/mod.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: tests/world.rs [hunk 15]
```diff
@@ -25,6 +25,15 @@ use common::unused_listener;
...
 use terminator::Terminator;
 
+fn build_small_queues<T: Send + 'static>() -> (PushQueues<T>, wireframe::push::PushHandle<T>) {
+    PushQueues::<T>::builder()
+        .high_capacity(1)
+        .low_capacity(1)
+        .rate(None)
+        .build()
+        .expect("failed to build PushQueues")
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/world.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: tests/wireframe_protocol.rs [hunk 16]
```diff
@@ -21,6 +21,23 @@ use wireframe::{
...
+// lint for single-line rstest fixtures on stable.
+#[allow(unfulfilled_lint_expectations)]
+fn queues() -> (PushQueues<Vec<u8>>, wireframe::push::PushHandle<Vec<u8>>) {
+    PushQueues::<Vec<u8>>::builder()
+        .high_capacity(8)
+        .low_capacity(8)
+        .rate(None)
+        .build()
+        .expect("failed to build PushQueues")
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/wireframe_protocol.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: docs/efficiency-report.md [hunk 17]
```diff
@@ -79,52 +91,62 @@ let mut out = Vec::with_capacity(self.0.len());
...
 - **Critical Path**: `LengthPrefixedProcessor::decode` method
-- **Optimization Priority**: High - affects every incoming frame
+- **Optimization Priority**: High—affects every incoming frame
 
 ### Connection Handling
 
 - **Bottleneck**: Connection actor event loop and fairness tracking
 - **Critical Path**: `tokio::select!` in connection actor
-- **Optimization Priority**: Medium - affects per-connection performance
+- **Optimization Priority**: Medium—affects per-connection performance
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/efficiency-report.md]: @coderabbitai Have the following now been resolved?

File: src/push/mod.rs [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+//! ```rust,no_run
+//! use wireframe::push::PushQueues;
+//!
+//! let (queues, handle) = PushQueues::<u8>::builder()
+//!     .high_capacity(8)
+//!     .low_capacity(8)
+//!     .build()
+//!     .expect("failed to build PushQueues");
+//! # drop((queues, handle));
+//! ```
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/mod.rs]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: src/push/queues/builder.rs [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+        self
+    }
+
+    /// Minimum interval between DLQ drop log entries.
+    #[must_use]
+    pub fn dlq_log_interval(mut self, interval: Duration) -> Self {
+        self.dlq_log_interval = interval;
+        self
+    }
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/queues/builder.rs]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: src/push/queues/errors.rs [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+    Closed,
+}
+
+/// Errors returned when creating push queues.
+#[non_exhaustive]
+#[derive(Debug, Error)]
+pub enum PushConfigError {
+    /// The provided rate was zero or exceeded [`MAX_PUSH_RATE`].
+    #[error("invalid rate {0}; must be between 1 and {max}", max = MAX_PUSH_RATE)]
+    InvalidRate(usize),
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/queues/errors.rs]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: src/push/queues/errors.rs [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+    /// The provided capacities were zero.
+    #[error("invalid capacities; high={high}, low={low}; each must be ≥ 1")]
+    InvalidCapacity { high: usize, low: usize },
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/queues/errors.rs]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: src/push/queues/handle.rs [hunk 22]
```diff
@@ -0,0 +1,248 @@
...
+                            log_interval = ?log_interval,
+                            "DLQ dropped frames (full or closed)"
+                        );
+                        *last = now;
+                        self.0.dlq_drops.store(0, Ordering::Relaxed);
+                    }
+                }
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/push/queues/handle.rs]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: tests/advanced/interaction_fuzz.rs [hunk 23]
```diff
@@ -23,7 +23,11 @@ enum Action {
 }
 
 async fn run_actions(actions: &[Action]) -> Vec<u8> {
-    let (queues, handle) = PushQueues::bounded(16, 16);
+    let (queues, handle) = PushQueues::<u8>::builder()
+        .high_capacity(16)
+        .low_capacity(16)
+        .build()
+        .expect("failed to build PushQueues");
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/advanced/interaction_fuzz.rs]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: tests/correlation_id.rs [hunk 24]
```diff
@@ -15,7 +15,12 @@ async fn stream_frames_carry_request_correlation_id() {
         yield Envelope::new(1, Some(cid), vec![1]);
         yield Envelope::new(1, Some(cid), vec![2]);
     });
-    let (queues, handle) = PushQueues::bounded(1, 1);
+    let (queues, handle) = PushQueues::<Envelope>::builder()
+        .high_capacity(1)
+        .low_capacity(1)
+        .rate(None)
+        .build()
+        .expect("failed to build PushQueues");
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/correlation_id.rs]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: tests/push_policies.rs [hunk 25]
```diff
@@ -73,11 +87,14 @@ fn push_policy_behaviour(
...
+fn dropped_frame_goes_to_dlq(rt: Runtime, builder: PushQueuesBuilder<u8>) {
     rt.block_on(async {
         let (dlq_tx, mut dlq_rx) = mpsc::channel(1);
-        let (mut queues, handle) = PushQueues::bounded_with_rate_dlq(1, 1, None, Some(dlq_tx))
-            .expect("queue creation failed");
+        let (mut queues, handle) = builder
+            .rate(None)
+            .dlq(Some(dlq_tx))
+            .build()
+            .expect("failed to build PushQueues");
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/push_policies.rs]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: tests/stream_end.rs [hunk 26]
```diff
@@ -2,19 +2,28 @@
...
 use terminator::Terminator;
 
+#[fixture]
+fn queues() -> (PushQueues<u8>, PushHandle<u8>) {
+    PushQueues::<u8>::builder()
+        .high_capacity(1)
+        .low_capacity(1)
+        .build()
+        .expect("failed to build PushQueues")
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/stream_end.rs]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.
